### PR TITLE
Revpi 2544/pb serdev

### PR DESCRIPTION
**Pull request by linosanfilippo-kunbus**

### CODE REVIEW

**Comment by l1k** [comment]
This <a href="https://github.com/RevolutionPi/linux/pull/112#issuecomment-1197913722">comment</a> has not been addressed AFAICS:

> Not sure if it makes sense to define the `current-speed` in the DT. The documentation in `Documentation/devicetree/bindings/serial/slave-device.txt` says:
> 
> > This should only be present in case a driver has no chance to know the baud rate of the slave device.
> >                   Examples:
> >                     * device supports auto-baud
> >                     * the rate is setup by a bootloader and there is no way to reset the device
> >                     * device baud rate is configured by its firmware but there is no way to request the actual settings
> 
> Normally the DT only describes things baked into the hardware. In this case, we know the baudrate is 115200 and we could set that directly in the driver. The driver could in theory negotiate a higher baudrate with devices if we'd amend their firmware to support it.

File: Documentation/devicetree/bindings/vendor-prefixes.yaml [hunk 1]
```diff
@@ -563,6 +563,8 @@ patternProperties:
     description: Kontron S&T AG
   "^kosagi,.*":
     description: Sutajio Ko-Usagi PTE Ltd.
+  "^kunbus,.*":
+    description: Kunbus GmbH
```
Comment (l1k) [inline Documentation/devicetree/bindings/vendor-prefixes.yaml]: Company name is KUNBUS GmbH (all caps) per www.handelsregister.de.

**Comment by l1k** [comment]
This <a href="https://github.com/RevolutionPi/linux/pull/112#discussion_r934274011">comment</a> on the compatible string is still unaddressed AFAICS. I've just responded to your reply, sorry for not doing this earlier.

**Review by l1k — CHANGES_REQUESTED** [review]
Aside from the three topics noted above this LGTM.

**Comment by linosanfilippo-kunbus** [comment]
> This [comment](https://github.com/RevolutionPi/linux/pull/112#discussion_r934274011) on the compatible string is still unaddressed AFAICS. I've just responded to your reply, sorry for not doing this earlier.

I changed the compatible string but chose "pi-bridge" instead of "pibridge" for now, since the latter is already used by the current piControl kernel module.

File: drivers/tty/serdev/pibridge.c [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+#define PIBRIDGE_RECV_BUFFER_SIZE	100
+#define PIBRIDGE_BC_ADDR		0xff
+
+#define PIBRIDGE_RESP_CMD		0x3fff
+#define PIBRIDGE_RESP_OK		0x4000
+#define PIBRIDGE_RESP_ERR		0x8000
+
+struct pibridge {
+	struct serdev_device *serdev;
+
```
Comment (nbuchwitz) [inline drivers/tty/serdev/pibridge.c]: Is this blank line intentional?

File: drivers/tty/serdev/pibridge.c [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+	u8	src;
+	u16	cmd;
+	u16	seq;
+	u8	len;
+} __attribute__((packed));
+
+struct pibridge_pkthdr_io {
+	u8 addr	:6;
+	u8 type	:1;	/* 0 for unicast, 1 for broadcast */
+	u8 rsp	:1;	/*always be 0 for sending, might be 1 for receiving*/
```
Comment (nbuchwitz) [inline drivers/tty/serdev/pibridge.c]: Looks like a missing space before and after the comment

File: arch/arm/boot/dts/overlays/revpi-connect-overlay.dts [hunk 4]
```diff
@@ -284,6 +284,9 @@
 			pinctrl-0 = <&rs485_pins>;
 			linux,rs485-enabled-at-boot-time;
 			status = "okay";
+			pi_bridge {
```
Comment (nbuchwitz) [inline arch/arm/boot/dts/overlays/revpi-connect-overlay.dts:288]: Missing new line here or one newline to much in the other overlay (line 262).  Sorry for nitpicking, just want to have this consistent  :smile:

File: drivers/tty/serdev/pibridge.c [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+
+	crc = pibridge_crc8(0, (u8 *) &pkthdr, sizeof(pkthdr));
+	crc = pibridge_crc8(crc, rcv_buf, rcv_len);
+
+	if (crc != crc_rcv) {
+		dev_warn_ratelimited(&pibridge_s->serdev->dev,
+			"check crc error in io-req\n");
+		return -EBADMSG;
+	}
+	/*received header check is not performed in io mode*/
```
Comment (nbuchwitz) [inline drivers/tty/serdev/pibridge.c]: Missing spaces before and after the comment

**Review by nbuchwitz — CHANGES_REQUESTED** [review]
Just some nitpicking, minor formatting issues

File: drivers/tty/serdev/pibridge.c [hunk 6]
```diff
@@ -108,6 +108,25 @@ static void pibridge_remove(struct serdev_device *serdev)
...
+
+	ret = wait_event_timeout(pi->read_queue,
+				 kfifo_len(&pi->read_fifo) >= len,
+				 msecs_to_jiffies(timeout));
+	mutex_lock(&pi->lock);
+	if (kfifo_len(&pi->read_fifo) >= len)
+		ret = 0;
+	else
+		ret = -1;
+	kfifo_reset(&pi->read_fifo);;
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c]: Nit: Duplicate semicolon.

File: drivers/tty/serdev/pibridge.c (hunk 3 above)
Comment (l1k) [reply drivers/tty/serdev/pibridge.c]: "might be" is a little confusing, it sounds like "perhaps it's 1 for responses but maybe it's not". I'd suggest something along the lines of: `/* 0 for request, 1 for response */`

File: drivers/tty/serdev/pibridge.c (hunk 2 above)
Comment (l1k) [reply drivers/tty/serdev/pibridge.c]: I guess sometimes a little grouping in structs doesn't hurt. ;)

File: drivers/tty/serdev/pibridge.c [hunk 7]
```diff
@@ -287,55 +290,64 @@ int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u16 snd_len,
 		return -EIO;
 	}
 
-	if (rcv_len != 0) {
-		if (pibridge_recv(rcv_buf, rcv_len) != rcv_len) {
+	crc = pibridge_crc8(0, (u8 *) &pkthdr, sizeof(pkthdr));
+
+	to_receive = min((u16) pkthdr.len, rcv_len);
+	to_discard = pkthdr.len - to_receive;
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:276]: Nifty, clean solution.

File: drivers/tty/serdev/pibridge.c [hunk 8]
```diff
@@ -287,55 +290,64 @@ int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u16 snd_len,
...
+		/* The provided buffer was too small. Discard the rest of the
+		   received data as well as the following CRC checksum byte. */
+		if (pibridge_discard_timeout(to_discard + 1, tmt))
+			dev_warn_ratelimited(&pibridge_s->serdev->dev,
+				"failed to discard %u bytes within timeout\n",
+				to_discard);
 		dev_warn_ratelimited(&pibridge_s->serdev->dev,
-			"receive crc error in gate-req\n");
-		return -EIO;
+			"received packet truncated (%u bytes missing)\n",
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:295]: I was confused at first that there's no `return` statement in the `if (to_discard)` branch, but I guess the CRC check below will fail (with high probability) for such packets, so maybe an explicit `return` indeed isn't necessary.

**Review by l1k — CHANGES_REQUESTED** [review]
With the bunch of nits addressed I think this will be good to go.

File: drivers/tty/serdev/pibridge.c (hunk 2 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c]: Yes, but in this case the blank line was unintended, so I will remove it.

File: drivers/tty/serdev/pibridge.c (hunk 8 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:295]: Note that in the discard case the CRC of the data portion is not built, so the CRC check wont fail. The caller wont get a negative return value but only a warning in the log. Should there rather be a negative return value?

Comment (l1k) [reply drivers/tty/serdev/pibridge.c:295]: Right after the `if ... else ...` statement there's a crc check:

```
	if (crc != crc_rcv)
		return -EBADMSG;
```

If the control flow entered the `if (to_discard)` branch, then `crc_rcv` is uninitialized and `crc` contains the CRC calculated over the header. With high probability, that check will fail and as a result, the function is exited with `-EBADMSG`.

An explicit `return` in the `if (to_discard)` branch might be cleaner.

If your intention is to not return an error, you need to change the control flow such that the crc check is skipped in the `if (to_discard)` case.

**Review by nbuchwitz — APPROVED** [review]
LGTM. Tested it successfully on arm32 and arm64.

**Comment by linosanfilippo-kunbus** [comment]
> LGTM. Tested it successfully on arm32 and arm64.

Great, thx for testing!

**Review by l1k — APPROVED** [review]

**Pull request merged**